lib/charms/loki_k8s/v0/loki_push_api.py

```diff
@@ -1183,8 +1186,11 @@ def __init__(
                 It is strongly advised not to change the default, so that people
                 deploying your charm will have a consistent experience with all
                 other charms that consume metrics endpoints.
-
+            port: the port on which Loki is running.
             rules_dir: path in workload container where rule files are to be stored.
+            scheme: whether Loki is running in http or https.
+            address: Loki server address.
```

Comment:
We are not documenting what happens with the default value as empty string, and we should because the behavior is non trivial.

Reply:
I think the default value for `address` should be `localhost`

Reply:
Absolutely not `localhost`. You are going to send `address` over the relation to a different container. The process there will think of its own `localhost`!

Reply:
No.
We are using `address` only to build the url used to check alert rules in `_check_alert_rules()` method:

```python
url = "{}://{}:{}{}/loki/api/v1/rules".format(self.scheme, self.address, self.port, path)
```

The endpoints we are sending over relation data come from ingress or are built on the fly, note that we are not using `address` here:

```python
    def _endpoints(self) -> List[dict]:
        """Return a list of Loki Push Api endpoints."""
        endpoints = []
        path = "/loki/api/v1/push"

        if self._charm.ingress_per_unit.urls:
            return [
                {"url": url.rstrip("/") + path}
                for url in self._charm.ingress_per_unit.urls.values()
            ]

        peers = [*self._charm.meta.peers.keys()][0]
        units = self._charm.model.get_relation(peers).units
        units.add(self._charm.unit)

        for unit in units:
            unit_number = unit.name.split("/")[-1]
            endpoints.append({"url": self._url(unit_number)})

        return endpoints

    def _url(self, unit_number) -> str:
        """Get the url for a given unit."""
        return "http://{}-{}.{}-endpoints.{}.svc.cluster.local:{}/loki/api/v1/push".format(
            self._charm.app.name,
            unit_number,
            self._charm.app.name,
            self._charm.model.name,
            self.port,
        )
```